src/VisualStudio/Core/Def/Implementation/TaskList/ProjectExternalErrorReporter.cs

```diff
@@ -25,37 +26,36 @@ internal class ProjectExternalErrorReporter : IVsReportExternalErrors, IVsLangua
         private readonly ProjectId _projectId;
         private readonly string _errorCodePrefix;
 
-        private readonly VisualStudioWorkspace _workspace;
-        private readonly ExternalErrorDiagnosticUpdateSource _diagnosticProvider;
+        private readonly VisualStudioWorkspaceImpl _workspace;
 
-        public ProjectExternalErrorReporter(ProjectId projectId, string errorCodePrefix, VisualStudioWorkspace workspace, ExternalErrorDiagnosticUpdateSource diagnosticProvider)
+        public ProjectExternalErrorReporter(ProjectId projectId, string errorCodePrefix, VisualStudioWorkspaceImpl workspace)
         {
             Debug.Assert(projectId != null);
             Debug.Assert(errorCodePrefix != null);
             Debug.Assert(workspace != null);
-            Debug.Assert(diagnosticProvider != null);
 
             _projectId = projectId;
             _errorCodePrefix = errorCodePrefix;
             _workspace = workspace;
-            _diagnosticProvider = diagnosticProvider;
 
             KnownUIContexts.SolutionBuildingContext.WhenActivated(() =>
             {
                 KnownUIContexts.SolutionBuildingContext.UIContextChanged += OnSolutionBuild;
-                _diagnosticProvider.OnSolutionBuildStarted();
+                DiagnosticProvider.OnSolutionBuildStarted();
             });
         }
 
+        private ExternalErrorDiagnosticUpdateSource DiagnosticProvider => _workspace.ExternalErrorDiagnosticUpdateSource;
```

Comment:
What did this buy? Would prefer if we had less tied to the workspace implementation guts if possible...

Reply:
this let external error diagnostic update source to be created more lazy. I guess I could leave it as MEF but change all caller to lazy<....>. but anyway, I did this way to make it lazy.